protostar/profiler/transaction_profiler.py

```diff
@@ -0,0 +1,226 @@
+# pylint: disable=invalid-name
+from dataclasses import dataclass, replace
+import itertools
+from typing import TYPE_CHECKING, List, Dict, Tuple
+
+from starkware.cairo.lang.tracer.tracer_data import TracerData
+from starkware.cairo.common.cairo_function_runner import CairoFunctionRunner
+from protostar.profiler.pprof import serialize, to_protobuf
+from protostar.profiler.contract_profiler import (
+    Instruction,
+    build_profile,
+    Function,
+    Sample,
+)
+
+
+if TYPE_CHECKING:
+    from protostar.starknet.cheatable_execute_entry_point import (
+        ContractProfile,
+        ContractFilename,
+    )
+
+Address = int
+GlobalFunctionID = str
+GlobalInstructionID = str
+
+GlobalFunction = Function
+GlobalInstruction = Instruction
+
+
+@dataclass(frozen=True)
+class TransactionProfile:
+    functions: List[GlobalFunction]
+    instructions: List[GlobalInstruction]
+    step_samples: List[Sample]
+    memhole_samples: List[Sample]
+
+
+def profile_from_tracer_data(tracer_data: TracerData, runner: CairoFunctionRunner):
+    assert runner.accessed_addresses
+    assert runner.segment_offsets
+    profile = build_profile(
+        tracer_data, runner.segments, runner.segment_offsets, runner.accessed_addresses
+    )
+
+    protobuf = to_protobuf(profile)
+    return serialize(protobuf)
+
+
+def translate_callstack(
+    current_contract: "ContractFilename",
+    in_instructions: Dict[Tuple[GlobalFunctionID, int], Instruction],
+    callstack: List[Instruction],
+) -> List[Instruction]:
+    new_callstack: List[Instruction] = []
+    for instr in callstack:
+        prefix = current_contract + "."
+        new_callstack.append(in_instructions[(prefix + instr.function.id, instr.id)])
+    return new_callstack
+
+
+def translate_callstacks(
+    current_contract: "ContractFilename",
+    global_instructions: Dict[Tuple[GlobalFunctionID, int], Instruction],
+    callstacks: List[List[Instruction]],
+) -> List[List[Instruction]]:
+    translated: List[List[Instruction]] = []
+    for call in callstacks:
+        translated.append(
+            translate_callstack(current_contract, global_instructions, call)
+        )
+    return translated
```

Comment:
Style: Consider top-down approach to vertical formatting, e.g.`translate_callstacks` should be above `translate_callstack`.

> Source code is like a newspaper article. The farther down we go, the more detailed the article gets.
https://levelup.gitconnected.com/javascript-clean-code-vertical-formatting-d15097f5f30e

Reply:
I am used to opposite approach bcs of c, I always write my code bottom up and I prefer it this way :p Why should decide on one format and use it globally in the project

Reply:
We need to define the style. It can be done here https://github.com/software-mansion/protostar/issues/928.

Reply:
I will change the ordering